drivers/counter/counter_sam_tc.c

```diff
@@ -0,0 +1,383 @@
+/*
+ * Copyright (c) 2021, Piotr Mienkowski
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT atmel_sam_tc
+
+/** @file
+ * @brief Atmel SAM MCU family counter (TC) driver.
+ *
+ * This version of the driver uses a single channel to provide a basic 16-bit
+ * counter. Remaining TC channels could be used in the future to provide
+ * additional functionality, e.g. input clock divider, 32-bit counter option,
+ * configured via DT properties.
+ *
+ * Remarks:
+ * - The driver is not thread safe.
+ * - The driver does not implement guard periods. Use at your own risk or
+ *   submit a patch.
+ */
+
+#include <errno.h>
+#include <sys/__assert.h>
+#include <sys/util.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <drivers/counter.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(counter_sam_tc, CONFIG_COUNTER_LOG_LEVEL);
+
+#define MAX_ALARMS_PER_TC_CHANNEL 2
+
+/* Device constant configuration parameters */
+struct counter_sam_dev_cfg {
+	struct counter_config_info info;
+	Tc *regs;
+	uint32_t reg_cmr;
+	void (*irq_config_func)(const struct device *dev);
+	const struct soc_gpio_pin *pin_list;
+	uint8_t pin_list_size;
+	uint8_t clk_sel;
+	uint8_t tc_chan_num;
+	uint8_t periph_id[TCCHANNEL_NUMBER];
+};
+
+struct counter_sam_alarm_data {
+	counter_alarm_callback_t callback;
+	void *user_data;
+};
+
+/* Device run time data */
+struct counter_sam_dev_data {
+	counter_top_callback_t top_cb;
+	void *top_user_data;
+
+	struct counter_sam_alarm_data alarm[MAX_ALARMS_PER_TC_CHANNEL];
+};
+
+#define DEV_NAME(dev) ((dev)->name)
+#define DEV_CFG(dev) \
+	((const struct counter_sam_dev_cfg *const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct counter_sam_dev_data *const)(dev)->data)
+
+static const uint32_t sam_tc_input_freq_table[] = {
+	USEC_PER_SEC,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 8,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 32,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 128,
+	32768,
+	USEC_PER_SEC, USEC_PER_SEC, USEC_PER_SEC,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ,
+};
+
+static int counter_sam_tc_start(const struct device *dev)
+{
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
```

Comment:
AFAICS, `TcChannel` is not available in other SAM series headers (i.e. `sam3*` and `sam4*`).